src/Gaffer/Animation.cpp

```diff
@@ -119,19 +119,23 @@ Animation::KeyPtr Animation::Key::setTime( float time )
 				//          insert clashing key into inactive keys container
 				//        else insert key into active keys container
 				// NOTE : It is critical to the following code that the key comparison NEVER throws.
-				assert( key->m_parent == curve );
+				// NOTE : key's parent, active or time state may have changed outside undo system in which case throw exception
+				if( key->m_parent != curve ) throw IECore::Exception( "Key parent changed outside undo system." );
+				if( key->m_active != active ) throw IECore::Exception( "Key active changed outside undo system." );
+				if( key->m_time != previousTime ) throw IECore::Exception( "Key time changed outside undo system." );
```

Comment:
Let's drop these checks into a little private `Key::throwIfStateCorrupted( curve, active, time )` method to avoid repeating ourselves here.